src/btree/bt_page.c

```diff
@@ -151,19 +151,22 @@ __page_inmem_prepare_update(WT_SESSION_IMPL *session, WT_ITEM *value, WT_CELL_UN
         WT_ERR(__wt_upd_alloc_tombstone(session, &tombstone, &size));
         total_size += size;
         tombstone->durable_ts = WT_TS_NONE;
-        tombstone->start_ts = unpack->tw.stop_ts;
         tombstone->txnid = unpack->tw.stop_txn;
         tombstone->prepare_state = WT_PREPARE_INPROGRESS;
+        tombstone->start_ts = unpack->tw.stop_ts;
+        tombstone->prepare_ts = unpack->tw.prepare_ts;
+        tombstone->prepared_id = unpack->tw.prepared_txn;
         F_SET(tombstone, WT_UPDATE_PREPARE_RESTORED_FROM_DS);
 
         /*
          * Mark the update also as in-progress if the update and tombstone are from same transaction
          * by comparing both the transaction and timestamps as the transaction information gets lost
          * after restart.
          */
-        if (unpack->tw.start_ts == unpack->tw.stop_ts &&
-          unpack->tw.durable_start_ts == unpack->tw.durable_stop_ts &&
-          unpack->tw.start_txn == unpack->tw.stop_txn) {
+        if ((unpack->tw.start_ts == unpack->tw.stop_ts &&
+              unpack->tw.durable_start_ts == unpack->tw.durable_stop_ts &&
+              unpack->tw.start_txn == unpack->tw.stop_txn) ||
+          WT_TIME_WINDOW_HAS_PREPARE(&unpack->tw)) {
```

Comment:
Why we need to check WT_TIME_WINDOW_HAS_PREPARE this now?

Reply:
We already know we need to reinstantiate a prepared update here. The question we try to answer here is that whether the prepare is the start, or the stop, or the both.